### PR TITLE
Add a hidden sequence point around the code that assigns pattern variables in switch

### LINKED ISSUE #14740: Unexpected stepping behavior in C# 7 switch statement
```break``` statements of cases whose condition is false are stepped onto like so:

image

That's confusing since ```break``` is not executed, the next step doesn't jump out of the switch statement but to the next case. Instead either the execution should first step on ```case int a``` then evaluate the condition to false and then step on the next case clause, or it should step directly to the body of the case that evaluates to true.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #14740

This is a hand-modified cherry-pick of the fix from dev15-rc

@AlekseyTs @VSadov Please re-review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs [hunk 1]
```diff
@@ -365,6 +365,41 @@ private void LowerConstantValueDecision(DecisionTree.ByValue byValue)
...
+            /// that binds the pattern variables.
+            /// </summary>
+            /// <param name="bindings">The source/destination pairs for the assignments</param>
+            private BoundStatement BindingsForCase(
+                ImmutableArray<KeyValuePair<BoundExpression, BoundExpression>> bindings)
+            {
+                var patternMatched = _factory.GenerateLabel("patternMatched");
+                _loweredDecisionTree.Add(_factory.Goto(patternMatched));
+
+                var addBindings = ArrayBuilder<BoundStatement>.GetInstance();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: >var addBindings = ArrayBuilder<BoundStatement>.GetInstance(); [](start = 16, length = 61)

It looks like we can simply pass ```ArrayBuilder<BoundStatement> sectionBuilder``` into this function and add directly into it.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs [hunk 2]
```diff
@@ -365,6 +365,41 @@ private void LowerConstantValueDecision(DecisionTree.ByValue byValue)
...
+                if (this._localRewriter.Instrument)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: >if (this._localRewriter.Instrument) [](start = 16, length = 35)

Let's add hidden sequence point unconditionally as we do in other places.

Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: We have mixed approaches, but I'm fine with making this one unconditional.

**Comment by gafter** [comment]
@AlekseyTs I've added a revision responsive to your comments.

**Comment by AlekseyTs** [comment]
LGTM. However, it looks like the change has a merge conflict.

**Comment by gafter** [comment]
Resolved merge with master.

@VSadov Please review.

**Comment by gafter** [comment]
@VSadov Please review

**Comment by VSadov** [comment]
LGTM

I think we need a bit more guidance on which sequence points are injected in the instrumenter and which are inserted in the course of lowering.
Intuitively it seems that lowering should only introduce hidden sequence points, but having a comment on the instrumenter explicitly saying so would be helpful.

Adding some clarification could be a separate issue from this. Unless you need to do more changes...

**Pull request merged**